### PR TITLE
Improve export logic

### PR DESCRIPTION
**Pull request by sakama**
Current implementation may cause OOM as @dmikurube commented at https://github.com/muga/embulk-output-elasticsearch/pull/32#discussion_r105589282.

I implemented original RecordBuffer to avoid this problem.
Also enabled unit test for export logic.

### CODE REVIEW

**Comment by sakama** [comment]
Fixed CI failure.

File: src/main/java/org/embulk/output/elasticsearch/ElasticsearchOutputPluginDelegate.java [hunk 1]
```diff
@@ -244,18 +243,16 @@ public ConfigDiff egestEmbulkData(final PluginTask task,
                                       int taskIndex,
                                       List<TaskReport> taskReports)
     {
-        ArrayNode records = JsonNodeFactory.instance.arrayNode();
+        int totalInserted = 0;
```
Comment (muga) [inline src/main/java/org/embulk/output/elasticsearch/ElasticsearchOutputPluginDelegate.java]: I think that long type is better.

File: src/main/java/org/embulk/output/elasticsearch/ElasticsearchOutputPluginDelegate.java [hunk 2]
```diff
@@ -244,18 +243,16 @@ public ConfigDiff egestEmbulkData(final PluginTask task,
...
         for (TaskReport taskReport : taskReports) {
-            if (taskReport.has("records")) {
-                records.addAll(JacksonTaskReportRecordBuffer.resumeFromTaskReport(taskReport, "records"));
+            if (taskReport.has("inserted")) {
+                totalInserted += taskReport.get(int.class, "inserted");
             }
         }
 
-        if (records.size() > 0) {
+        if (totalInserted > 0) {
```
Comment (muga) [inline src/main/java/org/embulk/output/elasticsearch/ElasticsearchOutputPluginDelegate.java]: If `totalInserted` is 0, we could skip index replacement?

File: src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            jacksonServiceRecord = (JacksonServiceRecord) serviceRecord;
+            JsonNode record = mapper.readTree(jacksonServiceRecord.toString()).get("record");
+
+            requestCount++;
+            totalCount++;
+            requestBytes += record.toString().getBytes().length;
+
+            records.add(record);
+            if (requestCount >= bulkActions || requestBytes >= bulkSize) {
+                client.push(record, task, retryHelper);
```
Comment (muga) [inline src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java]: record? records?

File: src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                if (totalCount % 10000 == 0) {
+                    log.info("Inserted {} total records", totalCount);
+                }
+                requestBytes = 0;
+                requestCount = 0;
```
Comment (muga) [inline src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java]: You don't need to clear records?

File: src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        catch (IOException ex) {
+            throw Throwables.propagate(ex);
+        }
+    }
+
+    @Override
+    public TaskReport commitWithTaskReportUpdated(TaskReport taskReport)
+    {
+        if (!records.toString().isEmpty()) {
+            client.push(records, task, retryHelper);
```
Comment (muga) [inline src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java]: record? records?

Comment (muga) [inline src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java]: Why records.toString is called? Why not use records.size?

File: src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+            requestCount++;
+            totalCount++;
+            requestBytes += record.toString().getBytes().length;
+
+            records.add(record);
+            if (requestCount >= bulkActions || requestBytes >= bulkSize) {
+                client.push(record, task, retryHelper);
+                if (totalCount % 10000 == 0) {
+                    log.info("Inserted {} total records", totalCount);
```
Comment (muga) [inline src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java]: I think that this is not "total".

**Review by muga — COMMENTED** [review]
The architecture looks good to me. I put several comments. Please check and fix them.

File: src/main/java/org/embulk/output/elasticsearch/ElasticsearchOutputPluginDelegate.java (hunk 1 above)
Comment (sakama) [reply src/main/java/org/embulk/output/elasticsearch/ElasticsearchOutputPluginDelegate.java]: Thanks.
Fixed at 8d936ae

File: src/main/java/org/embulk/output/elasticsearch/ElasticsearchOutputPluginDelegate.java (hunk 2 above)
Comment (sakama) [reply src/main/java/org/embulk/output/elasticsearch/ElasticsearchOutputPluginDelegate.java]: Ah right. we should replace index even if 0 record was inserted.
Fixed at 758c8ca

File: src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java (hunk 3 above)
Comment (sakama) [reply src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java]: Fixed at 8d936ae

File: src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java (hunk 6 above)
Comment (sakama) [reply src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java]: Fixed logging at 8d936ae

File: src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java (hunk 4 above)
Comment (sakama) [reply src/main/java/org/embulk/output/elasticsearch/ElasticsearchRecordBuffer.java]: Fixed at 8d936ae

**Comment by sakama** [comment]
Merged.
Please let me know if you have any concerns.

**Pull request merged**